flake: coderd: TestWorkspaceWatcher

Seen here: https://github.com/coder/coder/actions/runs/14194339815/job/39765912427

```
=== FAIL: coderd  (0.00s)
panic: test timed out after 10m0s
	running tests:
		TestWorkspaceWatcher (9m54s)
[...]
```

Logs: https://gist.githubusercontent.com/johnstcn/7cd8e954ca51bb8d84612b3825f1c31c/raw/e803449f445546d60d7de6e2709fe307ea0d33f8/gistfile1.txt

--- a/agent/agent.go
+++ b/agent/agent.go
@@ -1518,14 +1518,11 @@ func (a *agent) runCoordinator(ctx context.Context, tClient tailnetproto.DRPCTai
 	a.logger.Info(ctx, "connected to coordination RPC")
 
 	// This allows the Close() routine to wait for the coordinator to gracefully disconnect.
-	a.closeMutex.Lock()
-	if a.isClosed() {
-		return nil
+	disconnected := a.setCoordDisconnected()
+	if disconnected == nil {
+		return nil // already closed by something else
 	}
-	disconnected := make(chan struct{})
-	a.coordDisconnected = disconnected
 	defer close(disconnected)
-	a.closeMutex.Unlock()
 
 	ctrl := tailnet.NewAgentCoordinationController(a.logger, network)
 	coordination := ctrl.New(coordinate)
@@ -1545,6 +1542,17 @@ func (a *agent) runCoordinator(ctx context.Context, tClient tailnetproto.DRPCTai
 		}
 	}()
 	return <-errCh
+}
+
+func (a *agent) setCoordDisconnected() chan struct{} {
+	a.closeMutex.Lock()
+	defer a.closeMutex.Unlock()
+	if a.isClosed() {
+		return nil
+	}
+	disconnected := make(chan struct{})
+	a.coordDisconnected = disconnected
+	return disconnected
 }
 
 // runDERPMapSubscriber runs a coordinator and returns if a reconnect should occur.

